motoman_driver/src/io_relay.cpp

```diff
@@ -0,0 +1,127 @@
+ *  * Neither the name of the Southwest Research Institute, nor the names
+ *  of its contributors may be used to endorse or promote products derived
+ *  from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "motoman_driver/io_relay.h"
+#include <string>
+#include <ros/ros.h>
+
+using industrial::shared_types::shared_int;
+
+namespace motoman
+{
+namespace io_relay
+{
+
+// TODO: is this needed?
+#define ROS_ERROR_RETURN(rtn,...) do {ROS_ERROR(__VA_ARGS__); return(rtn);} while(0)
+
+bool MotomanIORelay::init(std::string default_ip, int default_port)
+{
+  ROS_INFO("MotomanIORelay: init");
+
+  std::string ip;
+  int port;
+
+  // override IP/port with ROS params, if available
+  ros::param::param<std::string>("robot_ip_address", ip, default_ip);
+  ros::param::param<int>("~port", port, default_port);
+
+  // check for valid parameter values
+  if (ip.empty())
+  {
+    ROS_ERROR("No valid robot IP address found.  Please set ROS 'robot_ip_address' param");
+    return false;
+  }
+
+  char* ip_addr = strdup(ip.c_str());  // connection.init() requires "char*", not "const char*"
+  ROS_INFO("I/O relay connecting to IP address: '%s:%d'", ip_addr, port);
+  default_tcp_connection_.init(ip_addr, port);
+  free(ip_addr);
+  default_tcp_connection_.makeConnect();
+
+  io_ctrl_.init(&default_tcp_connection_);
+  this->srv_read_single_io = this->node_.advertiseService("read_single_io",
+      &MotomanIORelay::readSingleIoCB, this);
+  this->srv_write_single_io = this->node_.advertiseService("write_single_io",
+      &MotomanIORelay::writeSingleIoCB, this);
+
+  return true;
+}
+
+// Service to read a single IO
+bool MotomanIORelay::readSingleIoCB(
+  motoman_msgs::ReadSingleIO::Request &req,
+  motoman_msgs::ReadSingleIO::Response &res)
+{
+  shared_int io_val= -1;
+
+  // send message and release mutex as soon as possible
+  this->mutex_.lock();
+  bool result = io_ctrl_.readSingleIO(req.address, io_val);
+  this->mutex_.unlock();
+
+  if (!result)
+  {
+    ROS_ERROR("Reading IO element %d failed", req.address);
+    return false;
+  }
+  res.value = io_val;
+  ROS_INFO("Element %d value: %d", req.address, io_val);
```

Comment:
Suggestion: make these `DEBUG` (same in `writeSingleIoCB(..)`).

Reply:
@gavanderhoorn Should I convert all DEBUG, ERROR, and INFO macros in this file to named versions? Or just the ones that you specified?

Also, should the name argument be the name of the class, `MotomanIORelay`?

Reply:
Would be nice to use `NAMED` everywhere, yes.

> Also, should the name argument be the name of the class, `MotomanIORelay`?

I wouldn't use the name of the class. We already know the node, the file and the line nr (via [this](http://wiki.ros.org/ROS/EnvironmentVariables#Console_Output_Formatting)), and the class doesn't add to much imo.

Suggestion: group log lines based on 'theme'. So in this case it would be `read`, earlier logs might be `init`, etc.

Reply:
Addressed in a6abe75

Sounds good. Let me know if any of these names should be changed.